### PR TITLE
Angular loss1.0

### PR DESCRIPTION
**Pull request by nithinraok**
- Added angular loss with cosine angle for 1.0
- Fixed multigpu metric issue by reusing classficationtopkaccuracy
- Added support for embedding extraction for speaker diarization

### CODE REVIEW

File: nemo/utils/exp_manager.py [hunk 1]
```diff
@@ -555,3 +561,12 @@ def on_train_end(self, trainer, pl_module):
...
     trainer.checkpoint_callback = checkpoint_callback
+
+
+def add_callbacks(trainer: 'pytorch_lightning.Trainer', nemo_callbacks: Optional[List[str]]):
+
+    for callback in nemo_callbacks:
+        if callback in callbacks.AVAILABLE_CALLBACKS:
+            trainer.callbacks.append(callbacks.AVAILABLE_CALLBACKS[callback])
+        else:
+            raise NameError(" Request callback is not part of nemo callbacks please check callback name")
```
Comment (blisc) [inline nemo/utils/exp_manager.py:572]: You really should look into hydra.instantiate. The solution works now, because these callbacks take no constructor arguments. But the moment you want to enhance your callback by saying moving `print_freq` to your callback constructor as opposed to taking it from trainer, this method will need to be improved.

Comment (nithinraok) [reply nemo/utils/exp_manager.py:572]: Agreed, Omry is working on a improved way of hydra instantiate, so will comeback on this when it was finalized 
viz: https://github.com/facebookresearch/hydra/issues/566

File: nemo/collections/asr/losses/angularloss.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import torch
+
+from nemo.core.classes import Loss, Typing, typecheck
+from nemo.core.neural_types import LabelsType, LogitsType, LossType, NeuralType
+from nemo.utils.decorators import experimental
+
+__all__ = ['AngularSoftmaxLoss']
+
+
+@experimental
```
Comment (blisc) [inline nemo/collections/asr/losses/angularloss.py:25]: Pleas drop experimental if it is ready.

**Review by blisc — COMMENTED** [review]
Mostly LGTM

File: nemo/utils/exp_manager.py (hunk 1 above)
Comment (blisc) [reply nemo/utils/exp_manager.py]: Moved out of this PR

File: nemo/collections/asr/losses/angularloss.py (hunk 2 above)
Comment (nithinraok) [reply nemo/collections/asr/losses/angularloss.py]: Done

File: examples/speaker_recognition/spkr_get_emb.py [hunk 3]
```diff
@@ -54,7 +54,6 @@ def main(cfg):
     model_path = os.path.join(log_dir, '..', 'spkr.nemo')
     speaker_model = ExtractSpeakerEmbeddingsModel.restore_from(model_path)
     speaker_model.setup_test_data(cfg.model.test_ds)
-
     trainer.test(speaker_model)
```
Comment (titu1994) [inline examples/speaker_recognition/spkr_get_emb.py:57]: Can you use the trainer.prepare_test() checks to ensure you dont eval on multiple gpus? Look at the speech_to_label.py script for usage.

File: nemo/collections/asr/data/audio_to_label.py [hunk 4]
```diff
@@ -134,24 +135,79 @@ def fixed_seq_collate_fn(self, batch):
...
+            batch (Optional[FloatTensor], Optional[LongTensor], LongTensor,
+                LongTensor):  A tuple of tuples of signal, signal lengths,
+                encoded tokens, and encoded tokens length.  This collate func
+                assumes the signals are 1d torch tensors (i.e. mono audio).
+            fixed_length (Optional[int]): length of input signal to be considered
+        """
+        slice_length = self.featurizer.sample_rate * self.time_length
+        _, audio_lengths, _, tokens_lengths = zip(*batch)
+        slice_length = min(slice_length, max(audio_lengths))
+        shift = 1 * 16000
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_label.py:170]: Hardcoded sample_rate? Replace with featurizer.sample_rate

File: nemo/collections/asr/losses/angularloss.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    @property
+    def output_types(self):
+        """Output types definitions for CTCLoss.
+        loss:
+            NeuralType(None)
+        """
+        return {"loss": NeuralType(elements_type=LossType())}
+
+    def __init__(self, s=20.0, m=1.35):
```
Comment (titu1994) [inline nemo/collections/asr/losses/angularloss.py:42]: No option to override epsilon for other tasks? Add default eps=1e-7.

Also, dont use 1 character names for variables. And add docstring to this class.

File: nemo/collections/asr/losses/angularloss.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        super().__init__()
+
+        self.eps = 1e-7
+        self.s = s
```
Comment (titu1994) [inline nemo/collections/asr/losses/angularloss.py:46]: Again, dont save variables with 1 char names. If its from a paper, add a reference section and explain what this variable is supposed to do (better yet, just use a descriptive name).

File: nemo/collections/asr/models/label_models.py [hunk 7]
```diff
@@ -50,7 +51,16 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         self.preprocessor = EncDecSpeakerLabelModel.from_config_dict(cfg.preprocessor)
         self.encoder = EncDecSpeakerLabelModel.from_config_dict(cfg.encoder)
         self.decoder = EncDecSpeakerLabelModel.from_config_dict(cfg.decoder)
-        self.loss = CELoss()
+        if 'angular' in cfg.decoder.params and cfg.decoder.params['angular']:
+            logging.info("Training with Angular Softmax Loss")
+            s = cfg.loss.s
```
Comment (titu1994) [inline nemo/collections/asr/models/label_models.py:56]: Config needs to have descriptive names, not one char variable name.

File: nemo/collections/asr/models/label_models.py [hunk 8]
```diff
@@ -139,38 +149,49 @@ def forward(self, input_signal, input_signal_length):
         return logits, embs
 
     # PTL-specific methods
-    def training_step(self, batch, batch_nb):
+    def training_step(self, batch, batch_idx):
+        # import ipdb; ipdb.set_trace()
```
Comment (titu1994) [inline nemo/collections/asr/models/label_models.py:153]: Remove comment strin

File: nemo/collections/asr/modules/conv_asr.py [hunk 9]
```diff
@@ -356,9 +356,21 @@ def output_types(self):
         )
 
     def __init__(
-        self, feat_in, num_classes, emb_sizes=[1024, 1024], pool_mode='xvector', init_mode="xavier_uniform",
+        self,
+        feat_in,
+        num_classes,
+        emb_sizes=[1024, 1024],
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:362]: Dont directly use lists here, use None and check for None below and create [1024, 1024] if None. Refer https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

File: nemo/collections/asr/modules/conv_asr.py [hunk 10]
```diff
@@ -399,9 +411,14 @@ def forward(self, encoder_output):
         embs = []
 
         for layer in self.emb_layers:
-            pool, emb = layer(pool), layer[:2](pool)
+            pool, emb = layer(pool), layer[: self.emb_id](pool)
             embs.append(emb)
 
+        if self.angular:
+            for W in self.final.parameters():
+                _ = F.normalize(W, p=2, dim=1)
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:419]: https://pytorch.org/docs/master/nn.functional.html#torch.nn.functional.normalize

F.normalize is not an inplace op unless you use `out=`, so whats the point of this loop then?

File: nemo/collections/common/callbacks/callbacks.py [hunk 11]
```diff
@@ -30,3 +32,40 @@ def on_epoch_end(self, trainer, pl_module):
...
+            print_freq = int(total_batches * print_freq)
+        if batch_idx % print_freq == 0:
+            logging.info(
+                "Epoch: {}/{} batch: {}/{} train_loss: {:.3f} train_acc: {:.2f}".format(
+                    trainer.current_epoch + 1,
+                    trainer.max_epochs,
+                    batch_idx + 1,
+                    total_batches,
+                    pl_module.loss_value,
+                    pl_module.accuracy,
```
Comment (titu1994) [inline nemo/collections/common/callbacks/callbacks.py:56]: Not all modules have `accuracy` so this callback will fail for a lot of models. Why not just read the log in its entirety and just print all of the values in the log? Cant we access the log at the end of train_batch_end?

File: nemo/collections/common/callbacks/callbacks.py [hunk 12]
```diff
@@ -30,3 +32,40 @@ def on_epoch_end(self, trainer, pl_module):
...
+
+
+class LogTrainValidLossCallback(Callback):
+    @rank_zero_only
+    def on_train_start(self, trainer, pl_module):
+        logging.info("Training started")
+
+    @rank_zero_only
+    def on_train_batch_end(self, trainer, pl_module, batch, batch_idx, dataloader_idx):
+        print_freq = trainer.row_log_interval
```
Comment (titu1994) [inline nemo/collections/common/callbacks/callbacks.py:44]: This will print every single batch (since PTL default is 10, nemo default is 1, not 1.0).

File: nemo/collections/common/callbacks/callbacks.py [hunk 13]
```diff
@@ -30,3 +32,40 @@ def on_epoch_end(self, trainer, pl_module):
...
+                    trainer.max_epochs,
+                    batch_idx + 1,
+                    total_batches,
+                    pl_module.loss_value,
+                    pl_module.accuracy,
+                )
+            )
+
+    def on_validation_epoch_end(self, trainer, pl_module):
+        logging.info(
```
Comment (titu1994) [inline nemo/collections/common/callbacks/callbacks.py:61]: Same, not all models have accuracy, so this will crash for them. is there a way to access the log dictionary itself?

**Review by titu1994 — CHANGES_REQUESTED** [review]
Some minor changes to pertinent to the model itself, and some major concerns regarding logging callbacks.

File: examples/speaker_recognition/spkr_get_emb.py (hunk 3 above)
Comment (nithinraok) [reply examples/speaker_recognition/spkr_get_emb.py:57]: I never faced issues with multi-GPU for extraction of my embeddings, I use it to gather my output. Is it necessary?

File: nemo/collections/asr/data/audio_to_label.py (hunk 4 above)
Comment (nithinraok) [reply nemo/collections/asr/data/audio_to_label.py:170]: Thanks missed it, Done

File: nemo/collections/asr/losses/angularloss.py (hunk 5 above)
Comment (nithinraok) [reply nemo/collections/asr/losses/angularloss.py:42]: eps is not a parameter, its to avoid negligible division by zero. Yes, I'll add docstring. s and m are very well known in angular loss literature for scale and margin. If it is compulsory I will look.

File: nemo/collections/asr/models/label_models.py (hunk 8 above)
Comment (nithinraok) [reply nemo/collections/asr/models/label_models.py:153]: Thanks, Done

File: nemo/collections/common/callbacks/callbacks.py (hunk 12 above)
Comment (nithinraok) [reply nemo/collections/common/callbacks/callbacks.py:44]: yes, its provided by user based on what % of num_batches or exact number he/she requires.

File: nemo/collections/asr/modules/conv_asr.py (hunk 9 above)
Comment (nithinraok) [reply nemo/collections/asr/modules/conv_asr.py:362]: Done

File: nemo/collections/asr/modules/conv_asr.py (hunk 10 above)
Comment (nithinraok) [reply nemo/collections/asr/modules/conv_asr.py:419]: It just normalizes the weights before calculating loss. I missed `W =` here, thanks Som

File: nemo/collections/common/callbacks/callbacks.py (hunk 11 above)
Comment (nithinraok) [reply nemo/collections/common/callbacks/callbacks.py:56]: Unfortunately not yet, PTL is working on it

File: examples/speaker_recognition/spkr_get_emb.py [hunk 14]
```diff
@@ -49,12 +39,15 @@
 def main(cfg):
 
     logging.info(f'Hydra config: {cfg.pretty()}')
-    trainer = pl.Trainer(logger=False, checkpoint_callback=False)
+    if cfg.trainer.gpus > 1:
```
Comment (titu1994) [inline examples/speaker_recognition/spkr_get_emb.py:42]: Wait do this only during inference (trainer.test()) otherwise you can't use multi GPU training

File: examples/speaker_recognition/spkr_get_emb.py (hunk 3 above)
Comment (titu1994) [reply examples/speaker_recognition/spkr_get_emb.py:51]: See above.

File: nemo/collections/asr/losses/angularloss.py [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+    @property
+    def output_types(self):
+        """Output types definitions for AngularLoss.
+        loss:
+            NeuralType(None)
+        """
+        return {"loss": NeuralType(elements_type=LossType())}
+
+    def __init__(self, scale=20.0, margin=1.35):
```
Comment (titu1994) [inline nemo/collections/asr/losses/angularloss.py:50]: Add default eps here inside Init.

File: nemo/collections/asr/modules/conv_asr.py [hunk 16]
```diff
@@ -356,12 +356,20 @@ def output_types(self):
...
+        self.emb_id = 2
+        if self.angular:
+            bias = False
+        else:
+            bias = True
 
         if type(emb_sizes) is str:
             emb_sizes = emb_sizes.split(',')
+        elif emb_sizes == None:
+            emb_sizes = [512, 512]
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:372]: Isn't this default different than the one above ? 512 instead of 1024

File: nemo/collections/asr/modules/conv_asr.py [hunk 17]
```diff
@@ -399,9 +407,14 @@ def forward(self, encoder_output):
         embs = []
 
         for layer in self.emb_layers:
-            pool, emb = layer(pool), layer[:2](pool)
+            pool, emb = layer(pool), layer[: self.emb_id](pool)
             embs.append(emb)
 
+        if self.angular:
+            for W in self.final.parameters():
+                W = F.normalize(W, p=2, dim=1)
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:415]: Have you trained a model and checked if you are getting expected output results ?

**Review by titu1994 — CHANGES_REQUESTED** [review]
Minor comments

File: examples/speaker_recognition/spkr_get_emb.py (hunk 14 above)
Comment (nithinraok) [reply examples/speaker_recognition/spkr_get_emb.py:42]: spkr_get_emb.py is only run for inference purposes.

File: nemo/collections/asr/modules/conv_asr.py (hunk 16 above)
Comment (nithinraok) [reply nemo/collections/asr/modules/conv_asr.py:372]: Doesn't matter

File: nemo/collections/asr/losses/angularloss.py (hunk 15 above)
Comment (nithinraok) [reply nemo/collections/asr/losses/angularloss.py:50]: eps is not a parameter to pass through init. Not required

File: nemo/collections/asr/modules/conv_asr.py (hunk 17 above)
Comment (nithinraok) [reply nemo/collections/asr/modules/conv_asr.py:415]: Yes, I was clearing LGTM errors didn't notice. This is corect

File: examples/speaker_recognition/spkr_get_emb.py (hunk 14 above)
Comment (titu1994) [reply examples/speaker_recognition/spkr_get_emb.py:42]: Ah ok.

**Review by titu1994 — COMMENTED** [review]
Looks good to go. I'll let @fayejf look it over for comments and then let's merge

**Review by titu1994 — APPROVED** [review]

File: nemo/collections/asr/data/audio_to_label.py [hunk 18]
```diff
@@ -134,24 +135,79 @@ def fixed_seq_collate_fn(self, batch):
                     sub = sig[-rem:] if rem > 0 else torch.tensor([])
                     rep_sig = torch.cat(repeat * [sig])
                     signal = torch.cat((rep_sig, sub))
+                    new_audio_lengths.append(torch.tensor(fixed_length))
```
Comment (fayejf) [inline nemo/collections/asr/data/audio_to_label.py:138]: what is the purpose for new_audio_length?

**Review by fayejf — APPROVED** [review]
Looks good to me! Just have two minor questions.

File: nemo/collections/asr/data/audio_to_label.py (hunk 18 above)
Comment (fayejf) [reply nemo/collections/asr/data/audio_to_label.py:138]: thanks for the explanation! Solved

File: examples/speaker_recognition/speaker_reco.py [hunk 19]
```diff
@@ -24,13 +24,14 @@
...
-        model.validation_ds.manifest_filepath="<validation_manifest_file>" \
-        hydra.run.dir="." \
-        trainer.gpus=0 \
-        trainer.max_epochs=50
-
+EXP_NAME=sample_run
+python ./speaker_reco.py --config-path='/mnt/ngc_ws/new_configs/' --config-name=$CONFIG_PATH \
+    trainer.max_epochs=10  \
+    model.train_ds.batch_size=64 model.validation_ds.batch_size=64 \
+    trainer.gpus=0 \
```
[the remark this answers is not in the document]
Comment (nithinraok) [reply examples/speaker_recognition/speaker_reco.py:31]: that section was for basic CPU run, I will add another section with hparams I used to converge the model.

[the remark this answers is not in the document]
Comment (fayejf) [reply examples/speaker_recognition/speaker_reco.py:31]: oh ok

**Comment by jainal09** [comment]
hey @nithinraok  i want to perform speaker diarialisation providing a audio file and getting multi speaker transcript ( stt of identified speakers) how to do that with this pr?

**Comment by nithinraok** [comment]
You could extract embeddings with [this](https://github.com/NVIDIA/NeMo/blob/main/examples/speaker_recognition/spkr_get_emb.py) script, and use those frame level embeddings to perform Spectral Clustering by mentioning num_speakers as number of clusters. We don't have this as a to go unified script for now, but all the individual pieces are already there. I will for sure added in next coming weeks.

**Comment by jainal09** [comment]
Thanks and waiting for these for a long time. @nithinraok
